Cj4 pfd touchup

charles-vomacka opened the pull request:
[PFD_Changes.pdf](https://github.com/Working-Title-MSFS-Mods/fspackages/files/6481451/PFD_Changes.pdf)
Changes explained in attached PDF document.

flymypi commented:
Great work & documentation and great find on some of those references!

A few notes:

- VNAV Constraint altitude
Good job on the sizing, could you move it about a third of a character to the right?
See the 3 of 3000 and 2 of 20000 positions in relation to the VS box.

- The font for the units in the baro preselect window needs to be bigger. I assume you have had to make it smaller because the font is monospaced and the dot in the inHg notation would cause overflow otherwise? 

- There's some versions of the PL21 software out there, so we'll have to check if that VS symbology fits the rest. Maybe @K20017 could evaluate that.

- This one is more of a request, but I've seen from your grid system taht our v-bar is too narrow at the moment. Would you be willing to adjust that, too? Otherwise I'd do it.

You can look for TheFlieger in our discord channel if you have questions.

charles-vomacka commented:
[VNAV moved.pdf](https://github.com/Working-Title-MSFS-Mods/fspackages/files/6487375/VNAV.moved.pdf)
Moved VNAV to correct position; see attached PDF.

Yes, the problem with the baro setting is the gap between the decimal point and the adjacent digits. I made the units font smaller so that it would fit in the preselect box. I considered breaking the setting into three parts; left, center, right. Then the gap could be removed from around the decimal. When in HPA mode, the entire thing could be placed into the right box. I was planning to try this as a further upgrade. I am new to web development (CSS/JS) and used this project as an exercise in learning, so do not know if this is possible, but will try to figure it out. 

I will see if I can figure out how to change the v-bar.

dga711 commented:
Awesome work @charles-vomacka . Thank you.
When all is done we would love to merge this for an upcoming CJ4 update close to SimUpdate 4.

flymypi commented:
> I considered breaking the setting into three parts; left, center, right. Then the gap could be removed from around the decimal. When in HPA mode, the entire thing could be placed into the right box. I was planning to try this as a further upgrade.

yes, that's the approach I would've suggested. You will have to stray into JS for that (and SVGs for the V-bars). Let me know if you need assistance or have any questions.

charles-vomacka commented:
I have discovered a bug when starting from cold and dark or when starting initialized on the runway - I had been doing all of my testing from a saved flight that was already airborne. There is an issue with the VS/VVS section of the FGSD - the vertical speedbug and direction arrow are displayed when they should not be. I will try to figure it out.

charles-vomacka commented:
Widened and flattened the v-bar to match reference.

![v-bar](https://user-images.githubusercontent.com/4685707/118371859-76e19700-b57c-11eb-8c92-463f3088b618.png)
![PFD_Ref_Distortion_Corrected](https://user-images.githubusercontent.com/4685707/118371899-b1e3ca80-b57c-11eb-92d8-4d78339acaf6.jpg)

charles-vomacka commented:
Moved center of aircraft symbol down slightly so that it does not clip v-bar as much.
![vbar2](https://user-images.githubusercontent.com/4685707/118372225-911c7480-b57e-11eb-9495-7fc96bc3ae10.png)

charles-vomacka commented:
Moved center of v-bar and aircraft symbol up slightly to correctly align them with the left and right rectangles. It looks like everything is correct now.
![vbar3](https://user-images.githubusercontent.com/4685707/118374591-fbd3ad00-b58a-11eb-8c58-f1578fd78a3c.png)
rectangles.
![PFD_Ref_Distortion_Corrected](https://user-images.githubusercontent.com/4685707/118374608-1ad23f00-b58b-11eb-8df5-1692da0a5bc7.png)

charles-vomacka commented:
I believe that I have successfully removed the gap between the decimal point and the adjacent digits in the altimeter setting. This allows the units portion to be sized correctly. This is my first Javascript, so would like someone to look at it before I commit. @flymypi  if I upload the file here, would you be able to take a look at it? I have a question about 3 particular lines.

![baroHG](https://user-images.githubusercontent.com/4685707/118418710-8c8fb300-b687-11eb-84cc-38f912d8ac51.png)
![baroHgstd](https://user-images.githubusercontent.com/4685707/118418717-91546700-b687-11eb-827f-55c7345ce2ae.png)
![baroHPA](https://user-images.githubusercontent.com/4685707/118418722-944f5780-b687-11eb-94d8-50334837882d.png)
![baroHPAstd](https://user-images.githubusercontent.com/4685707/118418724-96b1b180-b687-11eb-881c-a90b9adf3d52.png)
![baroPre02](https://user-images.githubusercontent.com/4685707/118418733-a0d3b000-b687-11eb-8071-cc4d35e2b72d.png)
![baroPre03](https://user-images.githubusercontent.com/4685707/118418734-a3360a00-b687-11eb-9176-388f7da751c3.png)

slip664 commented:
They usually tell me that the easiest way to review code is to just commit it to the branch. Also, make sure the new baro window sizes don't cover up the minimums that are displayed underneath it or you may have to move them around a little.
![1](https://user-images.githubusercontent.com/72432870/118433178-41799e00-b68f-11eb-8354-7db23fd7ca08.PNG)

charles-vomacka commented:
@slip Ok, thanks.

The window is the same size that it has always been; did not change it.

flymypi commented:
@charles-vomacka that looks great. Feel free to commit and comment the respective lines in this PR thread.

charles-vomacka commented on src/workingtitle-vcockpits-instruments-airliners/html_ui/Pages/VCockpit/Instruments/Airliners/Shared/WT/AltimeterIndicator.js:
```diff
@@ -107,8 +107,8 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
         }
-        if (!this.pressureSVG)
-            this.pressureSVG = document.createElementNS(Avionics.SVG.NS, "text");
+        if (!this.pressureSVGLeftPart)
+            this.pressureSVGLeftPart = document.createElementNS(Avionics.SVG.NS, "text");
```
@flymypi I was not sure what I should test for here; chose just to test for LeftPart. Do I need to test for all three parts (left, center, right)?

charles-vomacka replied:
Additionally, I think that there is a bug in this test in the current release. See issue #1160 I opened a while back.

flymypi replied:
Your fix for issue #1160 makes that entire check a bit redundant because you create that text-element anyway in line 181: `this.pressureSVGLeftPart = document.createElementNS(Avionics.SVG.NS, "text");`

Because it is now seperated into three parts and the construct should only run once, it's cleaner to just set create those elements before attaching the attributes. So I suggest you just remove this check.

charles-vomacka commented on src/workingtitle-vcockpits-instruments-airliners/html_ui/Pages/VCockpit/Instruments/Airliners/Shared/WT/AltimeterIndicator.js:
```diff
@@ -1551,39 +1602,59 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
         let baroPreset = SimVar.GetSimVarValue("L:XMLVAR_Baro1_SavedPressure", "number") / 16;
         
-        if (this.pressureSVG) {
+        if (this.pressureSVGLeftPart) {
```
@flymypi Again, not sure what to test for.

flymypi replied:
This check prevents the update loop tripping over a not yet constructed or "damaged" SVG structure. Because the following code depends on all three elements you should check for them.

```suggestion
        if (this.pressureSVGLeftPart && this.pressureSVGCenterPart && this.pressureSVGRightPart) {
```

charles-vomacka commented:
I think that I have completed the touch up, unless anyone wants anything else addressed.

flymypi commented on src/workingtitle-vcockpits-instruments-airliners/html_ui/Pages/VCockpit/Instruments/Airliners/Shared/WT/AltimeterIndicator.js:
```diff
@@ -1552,37 +1602,57 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
                     this.pressureBoxGroup.setAttribute("visibility", "visible");
-                    this.pressureSVG.removeAttribute("stroke", "black");
-                    this.pressureSVG.removeAttribute("stroke-width", "5px");
+                    this.pressureSVGLeftPart.removeAttribute("stroke", "black");
```
- I'm not sure what all this attribute adding and removing is for (I know it's been there before).

- Unless some parent element in the SVG defines it otherwise, the attribute `"stroke", "black"` is redundant.

I will check both of those notes in my sim.

charles-vomacka commented:
Removed check at line 110 as suggested. If this code is merged then issue #1160 can be closed.

charles-vomacka commented:
Updated check at line 1601 as suggested.

flymypi commented on src/workingtitle-vcockpits-instruments-airliners/html_ui/Pages/VCockpit/Instruments/Airliners/Shared/WT/AltimeterIndicator.js:
```diff
@@ -1551,39 +1600,59 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
-                    this.pressureSVG.removeAttribute("stroke", "black");
-                    this.pressureSVG.removeAttribute("stroke-width", "7px");
+                    this.pressureSVGLeftPart.removeAttribute("stroke", "black");
+                    this.pressureSVGLeftPart.removeAttribute("stroke-width", "5px");
+                    this.pressureSVGCenterPart.removeAttribute("stroke", "black");
+                    this.pressureSVGCenterPart.removeAttribute("stroke-width", "4px");
+                    this.pressureSVGRightPart.removeAttribute("stroke", "black");
+                    this.pressureSVGRightPart.removeAttribute("stroke-width", "5px");
                     this.pressureSVGUnits.removeAttribute("stroke", "black");
-                    this.pressureSVGUnits.removeAttribute("stroke-width", "7px");
+                    this.pressureSVGUnits.removeAttribute("stroke-width", "5px");
```
These attribute lines can be removed.

flymypi commented on src/workingtitle-vcockpits-instruments-airliners/html_ui/Pages/VCockpit/Instruments/Airliners/Shared/WT/AltimeterIndicator.js:
```diff
@@ -1551,39 +1600,59 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
-                this.pressureSVG.setAttribute("stroke", "black");
-                this.pressureSVG.setAttribute("stroke-width", "7px");
+                this.pressureSVGLeftPart.setAttribute("stroke", "black");
+                this.pressureSVGLeftPart.setAttribute("stroke-width", "5px");
+                this.pressureSVGCenterPart.setAttribute("stroke", "black");
+                this.pressureSVGCenterPart.setAttribute("stroke-width", "4px");
+                this.pressureSVGRightPart.setAttribute("stroke", "black");
+                this.pressureSVGRightPart.setAttribute("stroke-width", "5px");
                 this.pressureSVGUnits.setAttribute("stroke", "black");
-                this.pressureSVGUnits.setAttribute("stroke-width", "7px");
+                this.pressureSVGUnits.setAttribute("stroke-width", "5px");
```
These attribute lines can be removed.

charles-vomacka replied:
I think they are removed when the preset box is opened and then added back when closed. Since the stroke is black it probably would not show in the preset box anyway Should I remove this code and test what happens?

flymypi replied:
Yes, exactly. I have removed them on my end and didn't see any negative effect. You can verify that and commit.

charles-vomacka replied:
Lines removed.

flymypi commented on src/workingtitle-vcockpits-instruments-cj4/html_ui/Pages/VCockpit/Instruments/Airliners/CJ4/PFD/CJ4_PFD.html:
```diff
@@ -42,6 +42,11 @@
                             <div id="verticalCaptureDataField">
+                                <span id="verticalSpeedBugField_Icon">    
+                                    <svg id="verticalSpeedbug" viewBox="0 0 31 20">
+                                        <path d="M 20 0 L 32 10 L 20 20 L 20 15 L 15 15 L 15 5 L 20 5 L 20 0 Z"/>
```
```suggestion
                                        <path d="M 20 0 L 32 10 L 20 20 L 20 13 L 17 13 L 17 7 L 20 7 L 20 0 Z"/>
```

flymypi commented on src/workingtitle-vcockpits-instruments-airliners/html_ui/Pages/VCockpit/Instruments/Airliners/Shared/WT/AltimeterIndicator.js:
```diff
@@ -1606,15 +1606,6 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
-                    this.pressureSVGRightPart.removeAttribute("stroke-width", "5px");
-                    this.pressureSVGUnits.removeAttribute("stroke", "black");
-                    this.pressureSVGUnits.removeAttribute("stroke-width", "5px");
                     this._baroPresetChanged = baroPreset;
```
Those are still needed.

flymypi replied:
```suggestion
                    this.pressureBoxGroup.setAttribute("visibility", "visible");
                    this._baroPresetChanged = baroPreset;
```

charles-vomacka replied:
Sorry, was over zealous on my cut.

flymypi commented on src/workingtitle-vcockpits-instruments-airliners/html_ui/Pages/VCockpit/Instruments/Airliners/Shared/WT/AltimeterIndicator.js:
```diff
@@ -1639,20 +1630,10 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
-                this.pressureSVGRightPart.setAttribute("stroke-width", "5px");
-                this.pressureSVGUnits.setAttribute("stroke", "black");
-                this.pressureSVGUnits.setAttribute("stroke-width", "5px");
                 if (this.aircraft == Aircraft.A320_NEO) {
```
same here

flymypi replied:
```suggestion
                this.pressureBoxGroup.setAttribute("visibility", "hidden");
                if (this.aircraft == Aircraft.A320_NEO) {
```

charles-vomacka replied:
Sorry.

flymypi replied:
No problem at all, could've been clearer myself.
It looks like we're good to go now!

charles-vomacka commented:
@flymypi Ok. Thanks for your patience and help.

dga711 commented:
Again thank you for the contribution. Thats really great work!

charles-vomacka commented:
@dga711 No problem, I had fun doing it and learned new things.
